FLAMEGPU/FLAMEGPU2_dev
# General CMake Improvements

> Having reviewed 'Files changed', I'm comfortable with this pull request being squashed, so long as the full changelog is included within the message. The changes are quite clear from reading the changed files. 
> 
> This would correct the fact that half the commits fail to mention CMake.
>
> I think the compile warnings branch will survive (it's a branch of this). If not, I can cherrypick it into a new branch and fix conflicts. It mostly doesn't touch CMake (it only adds the forced warning stuff), so conflicts shouldn't be too bad.
> 
> Rob

This pull request significantly improves CMake, such that it works appropriately for both Visual Studio and g++. Hence, the `.sln` and `.vcxproj` files have been removed. Intuitive messages and options for users of both `cmake` and `cmake-gui` have been added. The readme has been updated to explain how to use CMake, on either platform, for those unfamiliar with it.

Closes #100 
Closes #96 

In future, when we have working examples, CMake can also be used to automatically configure the debugging settings (e.g. default working directory/args) within Visual Studio.

**Full Changelog:**
* Updates minimum cmake version to 3.12 (there was a comment saying <3.12 works with linker issues)
* Updates readme to provide platform agnostic `cmake` guidance.
* Adds delimiting message when individual projects are configured
* Unify warning level across C/CXX/CUDA
* Corrects handling of build configs between single/multi config generators.
* Reduces repeated messages when `common.cmake` is used by multiple targets.
* Set output dir OS specific (e.g. `linux-x64`, `windows-x64`)
* Fixes `SMS` variable to work via `cmake-gui`
* Disables generation of installation scripts
  * Hide `CMAKE_INSTALL_PREFIX` from `cmake-gui`
* Relocates binaries to within the build directory
  * `root-build-dir/bin/<os>/<config>`
  * Tests goto subdirectory `tests` of this path
  * `flamegpu2.lib` goes to `root-build-dir/lib/<os>/<config>`
  * This behaviour is mimicked if an example's `cmakelists.txt` is instead used.
* Renames `flamegpu2` sub-build dir from `src` to `FLAMEGPU2`
* Visual Studio:
  * Filters:
    * Builds visual studio filter hierarchy, mirroring directory structure
    * Groups targets into different filters
  * Reduce MSVC warning to W4
    * Suppress `unreferenced local function has been removed` warnings (on Windows)
  * Remove explicit Visual studio files and directories in favour of `cmake`
  * Sets startup project
    * Defaults to the top level target
    * For the root `CMakeLists.txt`, it sets to `flamegpu2` target.
* Lint:
  * Rename `lint` target to `all_lint`
  * Exclude lint targets from `ALL_BUILD` target
  * Bugfix: Only first file from list was being passed to cpplint
  * Prevents files in `external` dir from being linted
  * Adds check for `cpplint` to CMake
    * Gives warning if unavailable
    * Uses this path for calling cpplint (allows user to custom specify path if they so desire)
    * Doesn't create lint targets if cpplint not available
* CI:
  * Updates Travis to use recent Ubuntu, GCC, CUDA and CMake. 
  * Update Appveyor to use `cmake` (and CUDA 10)
  * Builds docs on Travis, but not appveyor *
* Docs:
  * Builds documentation using doxygen through cmake
  * Doxygen is an optional dependency - the docs target will only be generated if doxygen is found. 
* Misc:
  * Typos in comments
  * *Accidentally rebased in 9838db7 at some point 🤷‍♂️*
